### PR TITLE
Recover retention leases during peer recovery

### PR DESCRIPTION
**Pull request by jasontedor**
This commit integrates retention leases with recovery. With this change, we copy the current retention leases on primary to the replica during phase two of recovery. At this point, the replica will have been added to the replication group and so is already receiving retention lease sync requests from the primary. This means that if any retention lease syncs are triggered on the primary after we sample the retention leases here during phase two, that sync request will also arrive on the replica ensuring that the replica is from this point on up to date with the retention leases on the primary. We have to copy these during phase two since we will be applying indexing operations, potentially triggering merges, and therefore must ensure the correct retention leases are in place beforehand.

Relates #37165

### CODE REVIEW

**Comment by jasontedor** [comment]
@elasticmachine run elasticsearch-ci/default-distro

**Comment by jasontedor** [comment]
@elasticmachine run elasticsearch-ci/default-distro

**Review by dnhatn — APPROVED** [review]
I prefer not to sync RetentionLeases multiple times since syncing the leases once during recovery is enough. However, we can't do it without introducing a new step because we can't piggyback the leases in the prepareTranslog step nor the finalize step. Moreover, we expect to have a few leases, thus this choice makes a lot of sense to me.

File: server/src/main/java/org/elasticsearch/indices/recovery/RecoveryTranslogOperationsRequest.java [hunk 1]
```diff
@@ -39,18 +40,26 @@
     private int totalTranslogOps = RecoveryState.Translog.UNKNOWN;
     private long maxSeenAutoIdTimestampOnPrimary;
     private long maxSeqNoOfUpdatesOrDeletesOnPrimary;
+    private RetentionLeases retentionLeases;
```
Comment (dnhatn) [inline server/src/main/java/org/elasticsearch/indices/recovery/RecoveryTranslogOperationsRequest.java]: I think we need to initialize the retentionLeases with an empty leases in a mixed cluster?

**Comment by jasontedor** [comment]
> I prefer not to sync RetentionLeases multiple times since syncing the leases once during recovery is enough. However, we can't do it without introducing a new step because we can't piggyback the leases in the prepareTranslog step nor the finalize step. Moreover, we expect to have a few leases, thus this choice makes a lot of sense to me.

Indeed, I went through exactly the same dilemma. We *could* introduce a new step but I am not sure that it's worth it. Any thoughts @ywelsch before I push go on this?

File: server/src/main/java/org/elasticsearch/indices/recovery/RecoveryTranslogOperationsRequest.java (hunk 1 above)
Comment (jasontedor) [reply server/src/main/java/org/elasticsearch/indices/recovery/RecoveryTranslogOperationsRequest.java]: Great catch. I pushed 4d9fa1acddfe0b6e0ab8a91c10866ccbfa530e17.

**Comment by jasontedor** [comment]
@dnhatn and @ywelsch If we want a new step we can do it in a follow-up, I am going to keep moving for now.

**Pull request merged**